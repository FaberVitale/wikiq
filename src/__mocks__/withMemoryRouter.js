import React from "react";
import MemoryRouter from "react-router-dom/MemoryRouter";

export default (Comp) => (initialEntries = ["/"], initialIndex = 0) => {
  class MockRoute extends React.Component {
    static displayName = `withMockRoute(${
      Comp.displayName || Comp.name || "Component"
    })`;

    render() {
      return (
        <MemoryRouter
          initialEntries={initialEntries}
          initialIndex={initialIndex}
        >
          <Comp {...this.props} />
        </MemoryRouter>
      );
    }
  }
  return MockRoute;
};
