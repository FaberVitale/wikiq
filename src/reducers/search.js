//@flow
import {
  FETCH_SEARCH,
  STORE_SEARCH,
  FAILED_TO_FETCH_SEARCH,
  STORE_THUMBNAILS,
  FAILED_TO_FETCH_THUMBNAILS,
  FETCH_THUMBNAILS,
} from "../action/types";
import { PAGINATION_SIZE } from "../config";
import type { Action } from "../action/types";

export type WikiArticleInfo = {|
  +title: string,
  +description: string,
  +link: string,
|};

export type Thumbnail = {|
  +source: string,
  +width: number,
  +height: number,
|};

export type WikiArticle = {|
  +info: WikiArticleInfo,
  +thumbnail: Thumbnail | null,
|};

export type Search = {
  ids: Array<string> | null,
  fetchState: number,
  error: null | string,
  thumbnails: number,
};

export type Searches = {
  [id: string]: Search,
};

export type Articles = {
  [articleId: string]: WikiArticleInfo,
};

export type Thumbnails = {
  [articleId: string]: Thumbnail,
};

const DEFAULT_ERROR_MESSAGE = "error";

export const fetchStates = {
  IDLE: 0,
  REQUESTED_SEARCH: 1,
  REQUESTED_THUMBNAILS: 2,
};

Object.freeze(fetchStates);

const searchInitState = {
  ids: null,
  fetchState: fetchStates.IDLE,
  error: null,
  thumbnails: 0,
};

export const search = (state: Search = searchInitState, action: Action) => {
  switch (action.type) {
    case FETCH_SEARCH:
      return {
        ...state,
        fetchState: fetchStates.REQUESTED_SEARCH,
        error: null,
      };
    case STORE_SEARCH:
      return {
        ...state,
        ids: action.ids,
        fetchState: fetchStates.IDLE,
        error: null,
        thumbnails: PAGINATION_SIZE,
      };
    case FAILED_TO_FETCH_SEARCH:
      const message =
        action.error && typeof action.error.message === "string"
          ? action.error.message
          : DEFAULT_ERROR_MESSAGE;
      return {
        ...state,
        fetchState: fetchStates.IDLE,
        error: message,
      };
    case FETCH_THUMBNAILS:
      return {
        ...state,
        fetchState: fetchStates.REQUESTED_THUMBNAILS,
      };
    case STORE_THUMBNAILS:
    case FAILED_TO_FETCH_THUMBNAILS:
      return {
        ...state,
        thumbnails: state.thumbnails + PAGINATION_SIZE,
        fetchState: fetchStates.IDLE,
      };
    default:
      return state;
  }
};

export const searches = (state: Searches = {}, action: Action) => {
  switch (action.type) {
    case FAILED_TO_FETCH_SEARCH:
    case STORE_SEARCH:
    case FETCH_SEARCH:
    case STORE_THUMBNAILS:
    case FAILED_TO_FETCH_THUMBNAILS:
    case FETCH_THUMBNAILS:
      return {
        ...state,
        [action.searchId]: search(state[action.searchId], action),
      };
    default:
      return state;
  }
};

export const articles = (state: Articles = {}, action: Action) => {
  if (action.type === STORE_SEARCH) {
    const nextState = { ...state };

    const { ids, articles } = action;

    for (let i = 0, len = ids.length; i < len; i++) {
      nextState[ids[i]] = articles[i];
    }

    return nextState;
  }
  return state;
};

export const thumbnails = (state: Thumbnails = {}, action: Action) => {
  switch (action.type) {
    case STORE_SEARCH:
    case STORE_THUMBNAILS:
      return {
        ...state,
        ...action.thumbnails,
      };
    default:
      return state;
  }
};
