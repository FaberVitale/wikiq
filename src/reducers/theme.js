//@flow
import { CHANGE_THEME } from "../action/types";
import type { Action } from "../action/types";

export type State = "light" | "dark";

const initState: State = "light";

export default (state: State = initState, action: Action) => {
  if (action.type === CHANGE_THEME) {
    return state === "light" ? "dark" : "light";
  }
  return state;
};
