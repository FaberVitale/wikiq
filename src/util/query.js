//@flow
import type { WikiArticleInfo } from "../reducers";

// cache reference to the native methods
const baseEncodeURIComponent = encodeURIComponent;

const baseDecodeURIComponent = decodeURIComponent;

export const slugify = (val: string) =>
  baseEncodeURIComponent(val).replace(/%20/g, "_");

export const deslugify = (val: string) =>
  baseDecodeURIComponent(val.replace(/_/g, "\u0020"));

export const encodeComponent = (val: string) =>
  baseEncodeURIComponent(val).replace(/%20/g, "+");

export const decodeComponent = (val: string) =>
  baseDecodeURIComponent(val.replace(/\+/g, "%20"));

export const getOpenSearchURL = (
  lang: string,
  query: string,
  limit?: number = 10
) =>
  `https://${lang}.wikipedia.org/w/api.php?action=opensearch&search=${query}&limit=${limit}&namespace=0&format=json&origin=*`;

export const getThumbnailURL = (
  lang: string,
  titles: Array<string>,
  maxWidth: number = 100
) => {
  const query = titles.map(encodeURIComponent).join("|");

  return `https://${lang}.wikipedia.org/w/api.php?action=query&titles=${query}&prop=pageimages&format=json&origin=*&pithumbsize=${maxWidth}`;
};

export const makeSearchId = (lang: string, query: string) =>
  `/${lang}/${encodeComponent(query)}`;

/* Given a locale and a title slug returns the link to the pdf of the article*/
export const getPDFLink = (lang: string, title: string) =>
  `https://${lang}.wikipedia.org/api/rest_v1/page/pdf/${title}`;

export const getGoogleSearchLink = (query: string) =>
  `https://google.com/search?q=${encodeComponent(query.toLowerCase())}`;

export const fetchJSON: (
  input: RequestInfo,
  init?: RequestOptions
) => Promise<any> = async (input, init) => {
  const res = await fetch(input, init);

  if (res.status === 200) {
    return res.json();
  }

  let url = typeof input === "string" ? input : input.toString();

  throw new Error(
    `response status not 200. url: ${url}, status: ${res.status}`
  );
};

const makeArticleInfo: (
  title: string,
  description: string,
  link: string
) => WikiArticleInfo = (title, description, link) => ({
  title,
  description,
  link
});

export const makeArticleId = (lang: string, title: string) =>
  `${lang}/${encodeComponent(title)}`;

export const transformSearch: (
  lang: string,
  data: OpenSearch
) => { articles: Array<WikiArticleInfo>, ids: Array<string> } = (
  lang,
  data
) => {
  const [, titles, descriptions, links] = data;

  const articles = [];
  const ids = [];

  const len = titles.length;

  for (let i = 0; i < len; i++) {
    if (descriptions[i]) {
      articles.push(makeArticleInfo(titles[i], descriptions[i], links[i]));
      ids.push(makeArticleId(lang, titles[i]));
    }
  }

  return {
    articles,
    ids
  };
};

/* returns an object that maps titles to thumbnail links */
export const extractThumbnails = (lang: string, pages: {} = {}) => {
  const res = {};
  const ids = Object.keys(pages);

  for (let i = 0, len = ids.length, id = ids[i]; i < len; i++, id = ids[i]) {
    const title = pages[id].title;
    const thumbnail = pages[id].thumbnail;

    if (thumbnail) {
      res[makeArticleId(lang, title)] = thumbnail;
    }
  }

  return res;
};
