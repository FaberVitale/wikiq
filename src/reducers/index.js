//@flow
import { combineReducers } from "redux";
import theme from "./theme";
import type { State as Theme } from "./theme";
import { searches, articles, thumbnails } from "./search";
import type { Searches, Articles, Thumbnails } from "./search";

export { fetchStates } from "./search";

export type State = {|
  +theme: Theme,
  +searches: Searches,
  +articles: Articles,
  +thumbnails: Thumbnails
|};

export type { State as Theme } from "./theme";
export type { Search, Thumbnail, WikiArticle, WikiArticleInfo } from "./search";

export default combineReducers({
  theme,
  searches,
  articles,
  thumbnails
});
