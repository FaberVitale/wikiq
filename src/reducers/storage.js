//@flow
import type { Reducer } from "redux";
import { $localStorage } from "../util/dom";

/* higher order reducer that retrieves the initial state from the local storage
 * If the item is present is
 * transformed using a user-defined function onValue that defaults to JSON.parse,
 *
 * It is possible to validate the result passing a function that given an input
 *
 * signature:
 *  reducer: a valid reducer,
 *  key: used to request the initial State,
 *  options:
 *    storage: an object that has getItem function,
 *      defaults to localStore if present
 *    onValue: a function that given the string value recieved applies
 *      a transformation in order to deserialize the state,
 *      defaults to JSON.parse
 *    isValid: afunction that given the rehidrated value, validates it
 *      defaults to (val: S) => true
 */
interface GetItem {
  +getItem: (key: string) => ?string;
}

type Options<S> = {
  storage: GetItem,
  isValid: (value: S) => boolean,
  onValue: (value: string) => S,
};

const fauxStorage: GetItem = {
  getItem() {
    return null;
  },
};

const defOptions: Options<any> = {
  storage: $localStorage != null ? $localStorage : fauxStorage,
  isValid: (anyVal) => true,
  onValue: JSON.parse,
};
const storageType = "@@STORAGE_INIT";

export default <S, A>(
  reducer: Reducer<S, A>,
  key: string,
  options?: $Shape<Options<S>>
) => {
  const optionsMerged: Options<S> = Object.assign({}, defOptions, options);

  let storageValue: ?string = null;
  let initState;
  let hasAcquiredInitState = false;

  try {
    storageValue = optionsMerged.storage.getItem(key);
  } catch (e) {}

  if (storageValue != null) {
    try {
      let extractedValue = optionsMerged.onValue(storageValue);

      if (optionsMerged.isValid(extractedValue)) {
        hasAcquiredInitState = true;
        initState = extractedValue;
      }
    } catch (e) {}
  }

  if (!hasAcquiredInitState) {
    initState = reducer(undefined, { action: storageType });
  }

  return (state: S = initState, action: A) => reducer(state, action);
};
