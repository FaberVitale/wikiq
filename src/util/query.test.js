import * as fn from "./query";
import { openSearch, thumbnails } from "../__mocks__/api";
import { ids, articles, lang, query } from "../__mocks__/reducers";

describe("src/util/query", () => {
  const plus3 = "+++";
  const spaceSequence = "\u0020\u0020\u0020";
  const testString = "asg12AR#@~`\u0020/\\+()_:.,?'^ðŸ’©|&\"";

  test("encodeComponent", () => {
    expect(fn.encodeComponent(spaceSequence)).toBe(plus3);
    expect(
      decodeURIComponent(fn.encodeComponent(testString).replace(/\+/g, "%20"))
    ).toBe(testString);
  });

  test("decodeComponent", () => {
    expect(fn.decodeComponent(plus3)).toBe(spaceSequence);
    expect(
      fn.decodeComponent(encodeURIComponent(testString).replace(/%20/g, "%20"))
    ).toBe(testString);
  });

  test("slugify", () => {
    expect(fn.slugify(spaceSequence)).toBe("___");
    expect(
      decodeURIComponent(fn.slugify(testString).replace(/_/g, "%20"))
    ).toBe(testString.replace(/_/g, "\u0020"));
  });

  test("deslugify", () => {
    expect(fn.deslugify("___")).toBe(spaceSequence);
    expect(
      fn.deslugify(encodeURIComponent(testString).replace(/%20/g, "_"))
    ).toBe(testString.replace(/_/g, "\u0020"));
  });

  test("getOpenSearchURL", () => {
    expect(fn.getOpenSearchURL(lang, query)).toMatchSnapshot();
    expect(fn.getOpenSearchURL(lang, query, 30)).toMatchSnapshot();
  });

  test("getThumbnailURL", () => {
    const titles = ["as", "rem", "tik", "tak"];
    expect(fn.getThumbnailURL(lang, titles)).toMatchSnapshot();
    expect(fn.getThumbnailURL(lang, titles, 300)).toMatchSnapshot();
  });

  test("getPDFLink", () => {
    expect(fn.getPDFLink(lang, "stan")).toMatchSnapshot();
    expect(fn.getPDFLink(lang, "sting")).toMatchSnapshot();
  });

  test("transformSearch", () => {
    expect(fn.transformSearch(lang, openSearch)).toEqual({
      ids,
      articles: ids.map(id => articles[id])
    });
  });

  test("extractThumbnails", () => {
    const result = fn.extractThumbnails(lang, thumbnails.query.pages);
    const keys = Object.keys(result);

    expect(keys).not.toHaveLength(0);
    expect(
      keys.every(
        key =>
          Object.prototype.hasOwnProperty.call(result[key], "source") &&
          Object.prototype.hasOwnProperty.call(result[key], "width") &&
          Object.prototype.hasOwnProperty.call(result[key], "height")
      )
    ).toBe(true);
  });
});
