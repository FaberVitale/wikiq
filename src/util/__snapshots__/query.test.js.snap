// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/util/query getGoogleSearchLink 1`] = `"https://google.com/search?q=trip"`;

exports[`src/util/query getGoogleSearchLink 2`] = `"https://google.com/search?q=asg12ar%23%40~%60+%2F%5C%2B()_%3A.%2C%3F'%5E%F0%9F%92%A9%7C%26%22"`;

exports[`src/util/query getOpenSearchURL 1`] = `"https://en.wikipedia.org/w/api.php?action=opensearch&search=rem&limit=10&namespace=0&format=json&origin=*"`;

exports[`src/util/query getOpenSearchURL 2`] = `"https://en.wikipedia.org/w/api.php?action=opensearch&search=rem&limit=30&namespace=0&format=json&origin=*"`;

exports[`src/util/query getPDFLink 1`] = `"https://en.wikipedia.org/api/rest_v1/page/pdf/stan"`;

exports[`src/util/query getPDFLink 2`] = `"https://en.wikipedia.org/api/rest_v1/page/pdf/sting"`;

exports[`src/util/query getThumbnailURL 1`] = `"https://en.wikipedia.org/w/api.php?action=query&titles=as|rem|tik|tak&prop=pageimages&format=json&origin=*&pithumbsize=100"`;

exports[`src/util/query getThumbnailURL 2`] = `"https://en.wikipedia.org/w/api.php?action=query&titles=as|rem|tik|tak&prop=pageimages&format=json&origin=*&pithumbsize=300"`;
